Rewiring

evadelmas opened the pull request:
@TomClegg94 

- [X] ADBM rewiring 

- [X] Gilljam's rewiring 

- [ ] Model parameters: I am working on merging your `rewire_parameters` with the `model_parameter` function. Just to be sure, is your parameter `h` the same as the one from `model_parameter` (Hill coefficient for the functional response)? 

- [ ] Update `model_parameter` documentation 

- [ ] Staniczenko's rewiring

tpoisot committed:
Update simulate.jl

tpoisot committed:
Merge branch 'master' into tpoisot-patch-1

tpoisot committed:
Merge pull request #25 from PoisotLab/tpoisot-patch-1

Update simulate.jl

Commit:
Added rewiring

evadelmas committed:
Merge pull request #31 from TomClegg94/master

Gilljam and ADBM rewiring

tpoisot commented:
Just an update (for @TomClegg94 ) -- @evadelmas is working on adding the rewiring methods that are missing, and also updating the code to have an interface which is as simple as possible for the user. 

This will all happen on this pull request, so feel free to comment.

tpoisot committed:
:wrench: update Travis file

evadelmas committed:
merged rewire_parameters with make_parameters & updated make_parameters

evadelmas committed:
homogeneity : use of model_parameters instead of both model_parameters and rewire_parameters

evadelmas committed:
Fetch remote changes
Merge branch 'rewiring' of https://github.com/PoisotLab/BioEnergeticFoodWebs.jl into rewiring

CleggTom commented:
@evadelmas sounds good! 

The `:h` parameter I added is used in the ADBM to determine handling time as part of the optimal foraging model so we're gonna need to rename one of the terms. You can see its use the ADBM.jl file (in the rewiring folder) in the `getADBM_Terms` function. Sorry for any confusion this caused.

evadelmas commented:
@TomClegg94 Thanks! 
What about `h_adbm` and `a_adbm`? (we also have two `a` parameters)

CleggTom commented:
@evadelmas same as the `:h` parameter its used by the optimal foraging model in the ADBM

evadelmas commented:
@TomClegg94 Sorry my question was not clear at all. I have renamed the parameters `h`and `a` for the ADBM method `h_adbm` and `a_adbm`, are you ok with that or would you prefer something else?

CleggTom commented:
@evadelmas ah ok, that's good with me

evadelmas committed:
replaced p_r parameter in ABDM method

evadelmas committed:
Staniczenko's rewiring method

evadelmas committed:
sudo required

evadelmas committed:
fix isnan depwarn

evadelmas committed:
default to nonstiff

evadelmas committed:
fix depwarn

evadelmas committed:
include Staniczenko rewiring

evadelmas committed:
fix depwarn in test simulate

evadelmas committed:
fix depwarn in measures

evadelmas committed:
fix build

evadelmas committed:
fix build

evadelmas committed:
:rotating_light: model_parameters

evadelmas committed:
fetch change in travis file

evadelmas committed:
:rotating_light: empty extinctions vector when using Staniczenko's RM

evadelmas committed:
Update .travis.yml

evadelmas committed:
Fix depwarn

evadelmas committed:
Remove prints

evadelmas committed:
Change condition for ContinuousCallback to avoid monitoring extinct species

CleggTom commented:
Hi @evadelmas @tpoisot ,just wanted to check the progress.
Is there anything I can do to help?

evadelmas commented:
Hi @TomClegg94 
Sorry about the delay, I was working on some scholarship applications and did not work much on the package these past weeks.
To sum up: 
- all the code for the rewiring seems to be working, but
- there is a problem with the `ContinuousCallback` function, it seems that is does not always spot extinctions, especially when species' biomass drops very quickly in the first time steps (which is often the case). I have been working on that for a few days but did not manage to see what's wrong. 
- there is also a lot of tests to write with all the new code
I will start working on this tomorrow, if you have some time to help with any of it, please feel free to do so.

CleggTom commented:
Hey @evadelmas,

No worries, thats fine. good luck!

Thats weird with the callbacks. do you have any example cases?

I'll get on writing some tests, especially with the `ADBM` and `Gilljam` stuff

evadelmas commented:
Thanks for your help with the tests @TomClegg94 !

There is a reproducible example for the problem with the `simulate` function: 

```
A = [0 0 1; 0 1 0; 0 0 0]
rm = :Gilljam #same problem with any method
p = model_parameters(A, rewire_method = rm)
b = rand(size(p[:A],1))
s = simulate(p,b)
```

In this simple example, species 2 should disappear as it does not have any ressource beside itself, and species 1 and 3 should coexist. While 1 and 3 indeed persist, species 2 reaches a biomass ≈ 1e-70 but is not identified as an extinct species. 

```
p[:extinctions]
0-element Array{Int64,1}
```

```
s[:B][490:end,:]
12×3 Array{Float64,2}:
 0.0186646  1.70075e-68  0.0171995 
 0.015449   1.24243e-68  0.0380612 
 0.0146475  9.07631e-69  0.0841982 
 0.0176459  6.63036e-69  0.175522  
 0.0291579  4.84363e-69  0.316871  
 0.0636668  3.53839e-69  0.445694  
 0.153548   2.58487e-69  0.422244  
 0.290307   1.8883e-69   0.192125  
 0.344778   1.37944e-69  0.070876  
 0.309257   1.00771e-69  0.0260784 
 0.244737   7.36148e-70  0.00959587
 0.184244   5.3777e-70   0.00353135
```

tpoisot commented:
This callback issue is strange. We should have a more intense debugging session to make sure that there's not something weird on our end. If this is not giving any results, we can try reaching out to the DiffEqs maintainers to see what they think.

I guess tests come first, and when they are written we can get a better idea of where the problem may be.

tpoisot commented:
Looking at @evadelmas example, it seems that species 2 goes below 0 *during* the 1st step, so maybe it starts at 0 for the solver, and therefore is never removed? Maybe there is something in the callback code that happens when the change is immediate (as in the species never crosses the 0 threshold because it's never *above* 0)?

Also, what's the *definition* of 0 here? ˋ0.0ˋ ? ˋeps()ˋ ? ˋeps(0.0)ˋ ?

CleggTom commented:
I have a feeling it may be due to the disconnect between the callback condition (which triggers when any species biomasses crosses 0) and the setting of the derivative in the `dBdt` function:

```
  for i in eachindex(derivative)
    if dbdt[i] + biomass[i] < eps(0.0)
      derivative[i] = -biomass[i] 
    else
      derivative[i] = dbdt[i]
    end
  end
```

As it stands this causes the biomass of any species that was going to cross 0 to instead be set to 0.

edit: just as a quick test i changed the code above to: 

```
  for i in eachindex(derivative)
    if dbdt[i] + biomass[i] < eps(0.0)
      derivative[i] = -biomass[i] -0.1
    else
      derivative[i] = dbdt[i]
    end
  end
```
and its now detecting the extinctions.

tpoisot commented:
Good call. I forgot I added this before callbacks were added to DiffEqs.

evadelmas committed:
Remove condition on derivatives, add abstol and interp_points arg to the solve function, doesn't work

evadelmas commented:
Hi @TomClegg94,

A quick update on the situation: 
- We have tried implementing your solution (or more precisely, removing the `for i in  eachindex(derivative)` loop altogether as it's not needed anymore) but there are still some extinction we are not detecting. 
- We are now looking at new functionality from the Callback library (such as `PositiveDomain`) 

I am sorry it's taking so long, I hope we will be able to fix this soon. In the meantime, I'll keep working on the tests to make sure we can merge this branch as soon as the extinction problem is fixed.

ChrisRackauckas commented:
>This callback issue is strange. We should have a more intense debugging session to make sure that there's not something weird on our end. If this is not giving any results, we can try reaching out to the DiffEqs maintainers to see what they think.

I'd be glad to help.

>there is a problem with the ContinuousCallback function, it seems that is does not always spot extinctions, especially when species' biomass drops very quickly in the first time steps (which is often the case). I have been working on that for a few days but did not manage to see what's wrong.

I only quickly perused this, but did you take a look at this part of the docs:

http://docs.juliadiffeq.org/stable/features/callback_functions.html#Tweaking-the-Defaults-to-Specialized-Event-Detection-To-Your-Problem-1

If your problem isn't finding the event due to some kind of oscillatory behavior, you'll need to have more interpolation points. Or to detect a zero better for your case, you may want to adjust the `abstol` in the callback.

If I'm completely on the wrong page, please let me know and help me out a little bit since this is quite a big PR to dig into.

CleggTom commented:
@evadelmas no worries, do you have any code to recreate the errors?

evadelmas committed:
:bug: Return output even when all species are extinct

evadelmas committed:
:bug: Return output even when all species are extinct

evadelmas committed:
:art:

evadelmas commented:
Hi @ChrisRackauckas, 

Thanks a lot for offering to help.

Using the example described above, we have tried to modify the `ContinuousCallback` and `solve` functions defaults arguments, following what is described in the *"Tweaking the Defaults to Specialized Event Detection To Your Problem"* section of the documentation. 

More specifically we tried: 
- changing the limit to the timestep (`solve` with `dtmax = 10`, `dtmax = 100`, `dtmax = 1000`);
- increasing the `interp_points` up to `10000` in the `ContinuousCallback` function; 
- changing `abstol` to `1e-10` or `1e-5` in `ContinuousCallback`. 

Using all these different tweaks, we still have the same problem, with the extinction events detected or not, depending on the initial biomass. 

```
# Using this interaction matrix 
A = [0 0 1; 0 1 0; 0 0 0]

#this work:
p = model_parameters(A, rewire_method = :Gilljam)
b = [0.972 ; 0.323 ; 0.388]
s = simulate(p,b)
extinction time:2.9870307119376696 / sp.: 3 / b = -5.66618719598824e-23
extinction time:60.77982807719894 / sp.: 1 / b = 5.227781471335135e-22
extinction time:62.58615155864925 / sp.: 2 / b = 1.873544649899903e-29

#but using other values for the initial biomass vector often
#results in undetected extinctions:
p = model_parameters(A, rewire_method = :Gilljam)
b2 = [0.8 ; 0.5 ; 0.6]
s2 = simulate(p,b2)
s2

501×3 Array{Float64,2}:
 0.8         0.5          0.6        
 0.817872    0.30105      0.000231687
 0.597561    0.190523     2.47017e-7 
 0.43653     0.12534      2.40928e-9 
 0.318893    0.0848675    2.66996e-10
 0.232958    0.058682     3.56573e-11
 0.17018     0.0411955    1.04935e-11
 0.124319    0.0292375    5.4468e-12 
 0.090818    0.0209159    4.52564e-12
 0.0663439   0.0150482    5.15355e-12
 ⋮                                   
 0.00687112  1.96979e-68  0.0541019  
 0.00722602  1.43896e-68  0.126394   
 0.010488    1.05119e-68  0.265465   
 0.0217569   7.67922e-69  0.45709    
 0.0583525   5.60978e-69  0.597731   
 0.165514    4.09809e-69  0.528933   
 0.341823    2.99373e-69  0.179026   
 0.332534    2.18698e-69  0.0135252  
 0.249133    1.59763e-69  0.00148711 
``` 

I guess the problem might not related to the way we use the `solve` function but more about what's inside `ContinuousCallback`. Here is a quick description of the steps we use: 
- we create an empty vector to store the position of extinct species (`extspecies`) 
- we use `minimum(y[.!isext])` as condition for the callback to be sure not to keep tracking the biomass of species already extinct
- whenever this condition is true, the `affect!` function does the following:
	- update the parameters of the model 
	- add the position of the species identifies as extinct to `extspecies`
	- set the biomass of this species at `0` (`integrator.u[sp_min] = 0.0`) to avoid having extinct species coming back to life.

ChrisRackauckas committed:
fix callback condition

ChrisRackauckas commented:
It looks to me that I am getting 3 extinctions each time with both of them? I assume that stuff like

extinction time:1.8107702695177699 / sp.: 3 / b = -5.929230630780102e-21
extinction time:45.56205683817773 / sp.: 1 / b = 7.115076756936123e-20
extinction time:46.708878254954925 / sp.: 2 / b = -2.3781442610899546e-24

means 3 extinctions occurred, so that's good?

But maybe because that's on master I just fixed https://github.com/JuliaDiffEq/DifferentialEquations.jl/issues/215 which could be something that randomly showed up? Let me tag this version of OrdinaryDiffEq.jl with the fix and see if you get the same thing. If not, it seems like there's randomness here. If you could fix a random number seed to make it more easily reproducible that would help.

I'll ping when it merges.

ChrisRackauckas committed:
Fix diffeq setup

ChrisRackauckas committed:
install HDF5 on travis

evadelmas committed:
Merge pull request #35 from ChrisRackauckas/patch-2

install HDF5 on travis

evadelmas committed:
Merge branch 'master' into diffeq

evadelmas committed:
Merge pull request #33 from ChrisRackauckas/condition_fix

fix callback condition

ChrisRackauckas committed:
Fix type stability

evadelmas committed:
Merge pull request #36 from ChrisRackauckas/patch-2

Fix type stability

evadelmas committed:
info instead of println

evadelmas committed:
Merge + resolve conflicts

evadelmas committed:
Correction test simulate

evadelmas committed:
:white_check_mark: Species richness with total biomass = 0

evadelmas committed:
:white_check_mark: foodweb evenness

Commit:
Merge pull request #1 from PoisotLab/rewiring

Pulling from rewiring branch

Commit:
Merge pull request #2 from TomClegg94/rewiring

Rewiring

Commit:
Fixed :bug: with the getFeedingLinks where biomass
of 0 resulted in nan profits

Commit:
Added tests for ADBM and Gilljam rewiring

evadelmas committed:
Merge pull request #37 from TomClegg94/master

Fixing ADBM Bug and adding rewiring Tests

evadelmas committed:
:fire: Delete redundant makeParameters functions

evadelmas committed:
:fire: Remove export of rewire_parameters function

evadelmas committed:
:white_check_mark: save function

evadelmas committed:
:white_check_mark: Save as .json

evadelmas committed:
:white_check_mark: Update parameters

evadelmas committed:
Fix depwarn

evadelmas committed:
:white_check_mark: Update specialists preference

evadelmas committed:
Merge branch 'ChrisRackauckas-diffeq' into rewiring

evadelmas committed:
Remove arg. `derivative` from `dBdt()` as it causes fixed biomasses

ChrisRackauckas commented on src/dBdt.jl:
```diff
@@ -32,7 +32,7 @@ This function is the one wrapped by the various integration routines. Based on a
 parameters `p`, it will return `dB/dt` for every species.
 """
-function dBdt(t, biomass, p::Dict{Symbol,Any}, derivative)
+function dBdt(t, biomass, p::Dict{Symbol,Any})
```
what was the issue with the pre-allocated version?

evadelmas replied:
It prevents the biomasses from varying from one timestep to another, I have removed it for now until I can figure out what is happening. 
I have merged all your other changes but I still can't detect extinction events.

evadelmas committed:
Fix :bug: in Gilljam when the similarities vector is empty

evadelmas committed:
:fireworks: Simulation now detects extinctions :fireworks: (combination of tweaking biomasses in dbdt and changing abstol in the callback)

Commit:
Do not use callback if rewire_method = :none

Commit:
:white_check_mark: New tests for simulations.

ChrisRackauckas commented on src/dBdt.jl:
```diff
@@ -66,15 +66,14 @@ function dBdt(t, biomass, p::Dict{Symbol,Any})
-  #
-  # return derivative
+  for i in eachindex(dbdt)
+   if (dbdt[i] + biomass[i] < eps()) & (dbdt[i] + biomass[i] > 0.0)
```
I would probably put a tolerance on this, or at least make it more like 100eps(). `eps()` is small enough that you start getting numerical error when trying to hit it too closely.

ChrisRackauckas commented on src/simulate.jl:
```diff
@@ -81,7 +83,7 @@ function simulate(p, biomass; start::Int64=0, stop::Int64=500, use::Symbol=:nons
   end
 
-  sol = solve(prob, alg, callback = cb, saveat=t_keep, dense=false, save_timeseries=false)
+  sol = solve(prob, alg, callback = cb, dtmax = 1, saveat=t_keep, dense=false, save_timeseries=false)
```
Curious that this was found necessary. Did you try taking this off after the other checks were put on? Otherwise, if you'd like to give a test problem which needs it to me, I'd be happy to investigate what's going on. These floating point issues can be a real PITA.
